File: internal/policy/policy.go
package policy

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"

	"github.com/redtidev1918/releasegraph/internal/config"
	rgerrors "github.com/redtidev1918/releasegraph/internal/errors"
)

type Versioning struct {
	Mode     string `json:"mode" yaml:"mode"`
	Provider string `json:"provider,omitempty" yaml:"provider,omitempty"`
	Package  string `json:"package,omitempty" yaml:"package,omitempty"`
	Manifest string `json:"manifest,omitempty" yaml:"manifest,omitempty"`
	Version  string `json:"version,omitempty" yaml:"version,omitempty"`
}

type Tag struct {
	Template string `json:"template,omitempty" yaml:"template,omitempty"`
}

type BuildMatrixItem struct {
	Runner       string `json:"runner" yaml:"runner"`
	Command      string `json:"command" yaml:"command"`
	VersionCheck string `json:"version_check,omitempty" yaml:"versionCheck,omitempty"`
}

type Build struct {
	Test           string            `json:"test,omitempty" yaml:"test,omitempty"`
	Command        string            `json:"command,omitempty" yaml:"command,omitempty"`
	VersionCheck   string            `json:"version_check,omitempty" yaml:"versionCheck,omitempty"`
	FlutterVersion string            `json:"flutter_version,omitempty" yaml:"flutterVersion,omitempty"`
	Matrix         []BuildMatrixItem `json:"matrix,omitempty" yaml:"matrix,omitempty"`
}

type Assets struct {
	Required []string `json:"required" yaml:"required"`
	Optional []string `json:"optional" yaml:"optional"`
}

type Registry struct {
	Required bool   `json:"required" yaml:"required"`
	Publish  string `json:"publish,omitempty" yaml:"publish,omitempty"`
	Verify   string `json:"verify,omitempty" yaml:"verify,omitempty"`
	Context  string `json:"context,omitempty" yaml:"context,omitempty"`
	File     string `json:"file,omitempty" yaml:"file,omitempty"`
	Image    string `json:"image,omitempty" yaml:"image,omitempty"`
}

type Retention struct {
	Stable      int `json:"stable,omitempty" yaml:"stable,omitempty"`
	Prerelease  int `json:"prerelease,omitempty" yaml:"prerelease,omitempty"`
	FailedDraft int `json:"failed_draft,omitempty" yaml:"failed_draft,omitempty"`
}

type Release struct {
	Prerelease  bool   `json:"prerelease,omitempty" yaml:"prerelease,omitempty"`
	PostPublish string `json:"post_publish,omitempty" yaml:"post_publish,omitempty"`
	// GitHub declares whether a GitHub Release is part of this repository's
	// contract. nil means "not declared" and falls back to the historical
	// default (true), so existing policies keep their meaning.
	GitHub *bool `json:"github,omitempty" yaml:"github,omitempty"`
}

// Artifacts describes distributable build outputs. Repositories without
// distributable artifacts simply do not enable them.
type Artifacts struct {
	Binaries Binaries `json:"binaries,omitempty" yaml:"binaries,omitempty"`
}

// Binaries covers native/CLI artifacts. It is optional: a Python library, an npm
// package, a container service or a source-only repository has none.
type Binaries struct {
	Enabled *bool    `json:"enabled,omitempty" yaml:"enabled,omitempty"`
	Targets []string `json:"targets,omitempty" yaml:"targets,omitempty"`
}

type Policy struct {
	APIVersion string              `json:"apiVersion,omitempty" yaml:"apiVersion,omitempty"`
	Kind       string              `json:"kind" yaml:"kind"`
	Versioning Versioning          `json:"versioning" yaml:"versioning"`
	Tag        Tag                 `json:"tag,omitempty" yaml:"tag,omitempty"`
	Build      Build               `json:"build,omitempty" yaml:"build,omitempty"`
	Assets     Assets              `json:"assets" yaml:"assets"`
	Registries map[string]Registry `json:"registries,omitempty" yaml:"registries,omitempty"`
	Retention  Retention           `json:"retention,omitempty" yaml:"retention,omitempty"`
	Release    Release             `json:"release,omitempty" yaml:"release,omitempty"`
	Artifacts  Artifacts           `json:"artifacts,omitempty" yaml:"artifacts,omitempty"`
	Checksums  bool                `json:"checksums" yaml:"checksums"`
	Metadata   bool                `json:"metadata,omitempty" yaml:"metadata,omitempty"`
	SBOM       bool                `json:"sbom,omitempty" yaml:"sbom,omitempty"`
	Hash       string              `json:"hash,omitempty" yaml:"-"`
}

var versionPattern = regexp.MustCompile(`^[0-9]+(?:\.[0-9A-Za-z-]+)+$`)

var kinds = map[string]bool{"binary": true, "python-library": true, "node-library": true, "container": true, "flutter": true, "android": true, "hybrid": true, "none": true}
var versionModes = map[string]bool{"release-please": true, "manual": true, "tag": true}
var registryNames = map[string]bool{"github": true, "pypi": true, "npm": true, "pub": true, "ghcr": true}

func Load(path string) (*Policy, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, rgerrors.Wrap(rgerrors.Policy, "read policy", err)
	}
	return Parse(raw)
}

func Parse(raw []byte) (*Policy, error) {
	var p Policy
	if err := config.Unmarshal(raw, &p); err != nil {
		return nil, rgerrors.Wrap(rgerrors.Policy, "parse policy", err)
	}
	if p.APIVersion == "" {
		p.APIVersion = "releasegraph.dev/v1"
	}
	var rawMap map[string]any
	if err := yaml.Unmarshal(raw, &rawMap); err != nil {
		return nil, rgerrors.Wrap(rgerrors.Policy, "parse policy", err)
	}
	if _, ok := rawMap["checksums"]; !ok {
		p.Checksums = true
	}
	if p.Registries == nil {
		p.Registries = map[string]Registry{}
	}
	if err := Validate(&p); err != nil {
		return nil, err
	}
	sum := sha256.Sum256(raw)
	p.Hash = hex.EncodeToString(sum[:])
	return &p, nil
}

func Validate(p *Policy) error {
	if p.APIVersion != "" && p.APIVersion != "releasegraph.dev/v1" {
		return rgerrors.New(rgerrors.Policy, "apiVersion must be releasegraph.dev/v1")
	}
	if !kinds[p.Kind] {
		return rgerrors.New(rgerrors.Policy, "kind must be binary, python-library, node-library, container, flutter, android, hybrid, or none")
	}
	mode := p.Versioning.Mode
	if mode == "" {
		mode = p.Versioning.Provider
	}
	if !versionModes[mode] {
		return rgerrors.New(rgerrors.Policy, "versioning.mode must be release-please, manual, or tag")
	}
	if p.Tag.Template != "" && !strings.Contains(p.Tag.Template, "{version}") {
		return rgerrors.New(rgerrors.Policy, "tag.template must contain {version}")
	}
	for _, key := range []string{"required", "optional"} {
		values := map[string][]string{"required": p.Assets.Required, "optional": p.Assets.Optional}[key]
		for _, value := range values {
			if value == "" {
				return rgerrors.New(rgerrors.Policy, "assets."+key+" must contain non-empty strings")
			}
		}
	}
	names := make([]string, 0, len(p.Registries))
	for name := range p.Registries {
		names = append(names, name)
	}
	sort.Strings(names)
	for _, name := range names {
		if !registryNames[name] {
			return rgerrors.New(rgerrors.Policy, "unknown registry: "+name)
		}
		config := p.Registries[name]
		if containsNewline(config.Publish) || containsNewline(config.Verify) {
			return rgerrors.New(rgerrors.Policy, "registry publish and verify must be single-line commands")
		}
	}
	for _, item := range p.Build.Matrix {
		if item.Runner == "" || item.Command == "" {
			return rgerrors.New(rgerrors.Policy, "each build matrix item needs runner and command")
		}
		if containsNewline(item.Command) {
			return rgerrors.New(rgerrors.Policy, "build matrix commands must be single-line strings")
		}
	}
	if containsNewline(p.Release.PostPublish) {
		return rgerrors.New(rgerrors.Policy, "release.post_publish must be a single-line command")
	}
	return nil
}

func DesiredVersion(p *Policy, explicit, root string) (string, error) {
	if explicit != "" {
		version := trimPrefix(explicit, "v")
		if !versionPattern.MatchString(version) {
			return "", rgerrors.New(rgerrors.Policy, "invalid release version: "+explicit)
		}
		return version, nil
	}
	mode := p.Versioning.Mode
	if mode == "" {
		mode = p.Versioning.Provider
	}
	if mode == "manual" || mode == "tag" {
		if p.Versioning.Version == "" {
			return "", rgerrors.New(rgerrors.Policy, "manual versioning requires versioning.version or --version")
		}
		return DesiredVersion(&Policy{Versioning: Versioning{Mode: "manual", Version: p.Versioning.Version}}, p.Versioning.Version, root)
	}
	manifestPath := filepath.Join(root, defaultValue(p.Versioning.Manifest, ".release-please-manifest.json"))
	raw, err := os.ReadFile(manifestPath)
	if err != nil {
		return "", rgerrors.Wrap(rgerrors.Policy, "read release manifest", err)
	}
	return DesiredVersionFromManifest(p, raw)
}

func DesiredVersionFromManifest(p *Policy, raw []byte) (string, error) {
	var manifest map[string]string
	if err := json.Unmarshal(raw, &manifest); err != nil {
		return "", rgerrors.Wrap(rgerrors.Policy, "parse release manifest", err)
	}
	packageName := defaultValue(p.Versioning.Package, ".")
	version, ok := manifest[packageName]
	if !ok {
		return "", rgerrors.New(rgerrors.Policy, fmt.Sprintf("manifest has no package %q", packageName))
	}
	return DesiredVersion(&Policy{Versioning: Versioning{Mode: "manual", Version: version}}, version, ".")
}

func ReleaseTag(p *Policy, version string) string {
	return strings.ReplaceAll(defaultValue(p.Tag.Template, "v{version}"), "{version}", version)
}

func defaultValue(value, fallback string) string {
	if value != "" {
		return value
	}
	return fallback
}
func trimPrefix(value, prefix string) string {
	if len(value) >= len(prefix) && value[:len(prefix)] == prefix {
		return value[len(prefix):]
	}
	return value
}
func containsNewline(value string) bool {
	for i := range value {
		if value[i] == '\n' {
			return true
		}
	}
	return false
}

// Capabilities is the release contract of one repository, derived from its
// policy. ReleaseGraph is capability-driven: nothing here may be assumed for
// every repository.
type Capabilities struct {
	// GitHubRelease reports whether a public GitHub Release is part of the contract.
	GitHubRelease bool `json:"githubRelease"`
	// Binaries reports whether native artifacts are built and uploaded.
	Binaries bool `json:"binaries"`
	// Checksums reports whether SHA256SUMS must cover the required assets.
	Checksums bool `json:"checksums"`
	// Assets are the required asset patterns (may be empty: source-only repos).
	Assets []string `json:"assets,omitempty"`
	// Registries are the registry names that must be published.
	Registries []string `json:"registries,omitempty"`
}

// CapabilitiesOf derives the effective contract.
//
// Defaults preserve existing policies: a policy that builds a matrix or
// declares required assets has binaries; one that declares neither does not.
// Absence of a capability is a contract choice, never a release failure.
func (p *Policy) CapabilitiesOf() Capabilities {
	caps := Capabilities{GitHubRelease: true, Assets: append([]string{}, p.Assets.Required...)}

	if p.Release.GitHub != nil {
		caps.GitHubRelease = *p.Release.GitHub
	}
	switch {
	case p.Artifacts.Binaries.Enabled != nil:
		caps.Binaries = *p.Artifacts.Binaries.Enabled
	default:
		caps.Binaries = len(p.Build.Matrix) > 0 || len(p.Assets.Required) > 0
	}
	caps.Checksums = p.Checksums && len(caps.Assets) > 0

	names := make([]string, 0, len(p.Registries))
	for name, config := range p.Registries {
		if name == "github" || !config.Required {
			continue
		}
		names = append(names, name)
	}
	sort.Strings(names)
	caps.Registries = names
	return caps
}

// Names renders the capability set as stable identifiers used by ReleaseGraph
// compatibility metadata ("binary", "github-release", "checksums", registry
// names). Rollout compares these with a ReleaseGraph release's
// affected_capabilities to decide whether a repository is affected at all.
func (c Capabilities) Names() []string {
	names := []string{}
	if c.GitHubRelease {
		names = append(names, "github-release")
	}
	if c.Binaries {
		names = append(names, "binary")
	}
	if c.Checksums {
		names = append(names, "checksums")
	}
	for _, registry := range c.Registries {
		names = append(names, registry)
	}
	sort.Strings(names)
	return names
}

File: release_infra/policy.py
from __future__ import annotations

import hashlib
import json
import re
from pathlib import Path
from typing import Any


KINDS = {"binary", "python-library", "node-library", "flutter", "android", "container", "hybrid", "none"}
VERSIONING = {"release-please", "manual"}
REGISTRIES = {"github", "pypi", "npm", "pub", "ghcr"}


class PolicyError(ValueError):
    pass


def load_policy(path: str | Path = ".release-policy.yml") -> dict[str, Any]:
    raw = Path(path).read_bytes()
    try:
        policy = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise PolicyError("policy must use JSON syntax (valid YAML 1.2)") from exc
    validate_policy(policy)
    policy["_hash"] = hashlib.sha256(raw).hexdigest()
    return policy


# The asset pattern language.
#
# ReleaseGraph supports literal characters, `*` (any run, including empty) and
# `?` (exactly one character). Everything else is rejected.
#
# The reason is parity, and it is measured rather than assumed. Python's
# `fnmatch` and Go's `path.Match` are NOT the same language: `[!a]` means "not a"
# to fnmatch and "the literal characters ! and a" to path.Match, and `[^a]` means
# the exact opposite pair. CPython 3.14's fnmatch also compiles to atomic groups
# and lookaheads, which Go's RE2 cannot express at all, so porting it is
# impossible. Restricting the language to the region where both engines provably
# agree is what makes the two implementations comparable:
# `testdata/health/patterns.json` pins 484 (pattern, name) pairs that Python and
# Go are both required to match identically.
#
# The restriction costs nothing today: all 48 required patterns across the 16
# managed repositories use only literals and `*`, and none of the 137 released
# asset names contains a reserved character.
RESERVED_PATTERN_CHARS = "\\/[]!^"


def validate_asset_pattern(pattern: str) -> None:
    """Raise PolicyError unless the pattern is inside the supported language."""
    if not pattern:
        raise PolicyError("asset patterns must be non-empty")
    for char in pattern:
        if char in RESERVED_PATTERN_CHARS:
            raise PolicyError(
                f"asset pattern {pattern} uses reserved character {char}; "
                "supported syntax is literals, * and ?"
            )
        if ord(char) < 0x20 or ord(char) == 0x7F:
            raise PolicyError(f"asset pattern {pattern} contains a control character")


def validate_policy(policy: Any) -> None:
    if not isinstance(policy, dict):
        raise PolicyError("policy must be an object")
    if policy.get("kind") not in KINDS:
        raise PolicyError(f"kind must be one of: {', '.join(sorted(KINDS))}")
    versioning = policy.get("versioning")
    if not isinstance(versioning, dict) or versioning.get("mode") not in VERSIONING:
        raise PolicyError("versioning.mode must be release-please or manual")
    assets = policy.get("assets", {})
    for key in ("required", "optional"):
        if not isinstance(assets.get(key, []), list) or not all(isinstance(v, str) and v for v in assets.get(key, [])):
            raise PolicyError(f"assets.{key} must be a list of non-empty strings")
        for pattern in assets.get(key, []):
            validate_asset_pattern(pattern)
    registries = policy.get("registries", {})
    unknown = set(registries) - REGISTRIES
    if unknown:
        raise PolicyError(f"unknown registries: {', '.join(sorted(unknown))}")
    for name, config in registries.items():
        if not isinstance(config, dict) or config.get("required", True) not in (True, False):
            raise PolicyError(f"registries.{name} must be an object with boolean required")
        for command in ("publish", "verify"):
            value = config.get(command, "")
            if not isinstance(value, str) or "\n" in value:
                raise PolicyError(f"registries.{name}.{command} must be a single-line string")
    for name in ("test", "command", "version_check"):
        value = policy.get("build", {}).get(name, "")
        if not isinstance(value, str) or "\n" in value:
            raise PolicyError(f"build.{name} must be a single-line string")
    matrix = policy.get("build", {}).get("matrix")
    if matrix is not None:
        if not isinstance(matrix, list) or not matrix:
            raise PolicyError("build.matrix must be a non-empty array")
        allowed = {"runner", "command", "version_check"}
        for item in matrix:
            if not isinstance(item, dict) or not isinstance(item.get("runner"), str) or not item["runner"]:
                raise PolicyError("each build.matrix item needs a runner string")
            if not isinstance(item.get("command"), str) or "\n" in item["command"]:
                raise PolicyError("each build.matrix item needs a single-line command")
            if set(item) - allowed:
                raise PolicyError(f"unknown build.matrix field(s): {', '.join(sorted(set(item) - allowed))}")
    post_publish = policy.get("release", {}).get("post_publish", "")
    if not isinstance(post_publish, str) or "\n" in post_publish:
        raise PolicyError("release.post_publish must be a single-line string")


def desired_version(policy: dict[str, Any], explicit: str | None = None, root: str | Path = ".") -> str:
    if explicit:
        version = explicit.removeprefix("v")
        if not re.fullmatch(r"[0-9]+(?:\.[0-9A-Za-z-]+)+", version):
            raise PolicyError(f"invalid release version: {explicit}")
        return version
    versioning = policy["versioning"]
    if versioning["mode"] == "manual":
        version = versioning.get("version")
        if not version:
            raise PolicyError("manual versioning requires versioning.version or --version")
        return desired_version({"versioning": {"mode": "manual", "version": None}}, str(version))
    manifest_path = Path(root) / versioning.get("manifest", ".release-please-manifest.json")
    manifest = json.loads(manifest_path.read_text())
    package = versioning.get("package", ".")
    if package not in manifest:
        raise PolicyError(f"manifest has no package {package!r}")
    return desired_version({"versioning": {"mode": "manual", "version": None}}, str(manifest[package]))
